File: cli-go/internal/proposal/clocks_test.go
package proposal

import (
	"testing"
	"time"
)

// iso renders an epoch-ms as the wire timestamp (Date.prototype.toISOString).
func iso(ms int64) string {
	return time.UnixMilli(ms).UTC().Format("2006-01-02T15:04:05.000Z07:00")
}

const day = int64(24 * 60 * 60 * 1000)

func TestFormatDuration(t *testing.T) {
	cases := []struct {
		ms   int64
		want string
	}{
		{0, "0s"},
		{5_000, "5s"},
		{-5_000, "5s"}, // abs
		{90_000, "1m 30s"},
		{9 * 60_000, "9m 0s"},
		{10 * 60_000, "10m"},
		{65 * 60_000, "1h 5m"},
		{(26*3600 + 5*60) * 1000, "1d 2h"},
	}
	for _, c := range cases {
		if got := FormatDuration(c.ms); got != c.want {
			t.Errorf("FormatDuration(%d) = %q, want %q", c.ms, got, c.want)
		}
	}
}

// TestClocksStayDistinct is the load-bearing two-clock test (s07 §T0): a row that
// carries BOTH expiresAt and holdUntil must never render them as one number. A
// pending row shows expiresIn (from expiresAt) and no hold; a held row shows
// holdRemaining (from holdUntil) and no expiry — each sourced from its own field,
// gated on status. Mirrors clocks.test.ts's "row carrying both".
func TestClocksStayDistinct(t *testing.T) {
	now := time.Now().UnixMilli()

	pending := &Proposal{
		Status:    "pending",
		CreatedAt: iso(now - 2*day),
		ExpiresAt: iso(now + 3*day), // decision deadline
		HoldUntil: iso(now + 1*day), // present but MUST be ignored while pending
	}
	pc := pending.Clocks(now)
	if pc.ExpiresInMs == nil {
		t.Fatal("pending: expiresIn must be set from expiresAt")
	}
	if pc.HoldRemainingMs != nil {
		t.Error("pending: holdRemaining must be nil — holdUntil is not this row's clock")
	}
	if got := *pc.ExpiresInMs; got < 2*day || got > 3*day {
		t.Errorf("pending: expiresIn = %d, want ~3d", got)
	}

	held := &Proposal{
		Status:    "held",
		CreatedAt: iso(now - 2*day),
		DecidedAt: iso(now - 1*day),
		ExpiresAt: iso(now + 3*day), // present but MUST be ignored while held
		HoldUntil: iso(now + 4*60_000),
	}
	hc := held.Clocks(now)
	if hc.HoldRemainingMs == nil {
		t.Fatal("held: holdRemaining must be set from holdUntil")
	}
	if hc.ExpiresInMs != nil {
		t.Error("held: expiresIn must be nil — the deadline is moot once held")
	}
	if got := *hc.HoldRemainingMs; got <= 0 || got > 5*60_000 {
		t.Errorf("held: holdRemaining = %d, want ~4m", got)
	}
}

func TestWaitedForMs_FreezesOnDecision(t *testing.T) {
	now := time.Now().UnixMilli()

	// Pending: still accruing → now - created.
	p := &Proposal{Status: "pending", CreatedAt: iso(now - 3*day)}
	if got := p.WaitedForMs(now); got < 3*day-1000 || got > 3*day+1000 {
		t.Errorf("pending waited = %d, want ~3d", got)
	}
	// Decided: frozen at decidedAt (history must not keep accruing shame).
	d := &Proposal{Status: "approved", CreatedAt: iso(now - 3*day), DecidedAt: iso(now - 1*day)}
	if got := d.WaitedForMs(now); got != 2*day {
		t.Errorf("decided waited = %d, want exactly 2d", got)
	}
	// Expired: no decidedAt, freezes at expiresAt (the sweep writes no decidedAt).
	e := &Proposal{Status: "expired", CreatedAt: iso(now - 3*day), ExpiresAt: iso(now - 1*day)}
	if got := e.WaitedForMs(now); got != 2*day {
		t.Errorf("expired waited = %d, want exactly 2d", got)
	}
}

func TestLabels(t *testing.T) {
	overdue := int64(-90_000)
	future := int64(2 * 60 * 60 * 1000)
	var none *int64
	holdFuture := int64(3 * 60_000)
	holdLapsed := int64(-1)

	if got := ExpiryLabel(&future); got != "expires in 2h 0m" {
		t.Errorf("ExpiryLabel(future) = %q", got)
	}
	if got := ExpiryLabel(&overdue); got != "decision window passed 1m 30s ago" {
		t.Errorf("ExpiryLabel(overdue) = %q", got)
	}
	if got := ExpiryLabel(none); got != "no deadline" {
		t.Errorf("ExpiryLabel(nil) = %q", got)
	}
	// The hold label must never read as egress.
	if got := HoldLabel(&holdFuture); got != "retraction window ends in 3m 0s — nothing has been sent" {
		t.Errorf("HoldLabel(future) = %q", got)
	}
	if got := HoldLabel(&holdLapsed); got == "" || !contains(got, "still not sent") {
		t.Errorf("HoldLabel(lapsed) = %q, must say still-not-sent", got)
	}
	if got := HoldLabel(nil); got != "" {
		t.Errorf("HoldLabel(nil) = %q, want empty", got)
	}
	pendingWaited := WaitedLabel("pending", RowClocks{WaitedMs: 5000})
	if pendingWaited != "waiting on you 5s" {
		t.Errorf("WaitedLabel(pending) = %q", pendingWaited)
	}
	if got := WaitedLabel("approved", RowClocks{WaitedMs: 5000}); got != "sat with you 5s" {
		t.Errorf("WaitedLabel(decided) = %q", got)
	}
}

// TestOrderQueue pins the urgency order: pending soonest-deadline-first (no
// deadline last), then held, then history newest-first (clocks.ts:138).
func TestOrderQueue(t *testing.T) {
	now := time.Now().UnixMilli()
	// Deliberately NOT in order, so a caller that forgot to sort fails.
	in := []*Proposal{
		{ID: "history", Status: "approved", DecidedAt: iso(now - 10*day), CreatedAt: iso(now - 11*day)},
		{ID: "no-deadline", Status: "pending", CreatedAt: iso(now - 3*day)},
		{ID: "held", Status: "held", HoldUntil: iso(now + 60_000), CreatedAt: iso(now - 5*day)},
		{ID: "later", Status: "pending", ExpiresAt: iso(now + 2*day), CreatedAt: iso(now - 1*day)},
		{ID: "soon", Status: "pending", ExpiresAt: iso(now + 1*day), CreatedAt: iso(now - 1*day)},
	}
	got := OrderQueue(in)
	want := []string{"soon", "later", "no-deadline", "held", "history"}
	for i, w := range want {
		if got[i].ID != w {
			t.Fatalf("OrderQueue position %d = %q, want %q (full: %s)", i, got[i].ID, w, ids(got))
		}
	}
}

func ids(ps []*Proposal) string {
	out := ""
	for i, p := range ps {
		if i > 0 {
			out += ","
		}
		out += p.ID
	}
	return out
}

func contains(s, sub string) bool {
	return len(sub) == 0 || (len(s) >= len(sub) && indexOf(s, sub) >= 0)
}

func indexOf(s, sub string) int {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return i
		}
	}
	return -1
}

File: webmail/src/lib/approvals/clocks.test.ts
import { describe, expect, it } from "vitest";
import {
  NEAR_EXPIRY_MS,
  expiryLabel,
  formatDuration,
  holdLabel,
  isNearExpiry,
  orderQueue,
  rowClocks,
  waitedForMs,
  waitedLabel,
} from "./clocks";
import type { ActionProposal } from "./types";

const NOW = Date.parse("2026-08-11T12:00:00Z");
const MIN = 60_000;
const HOUR = 3600_000;
const DAY = 24 * HOUR;

function proposal(partial: Partial<ActionProposal> & Pick<ActionProposal, "id">): ActionProposal {
  return {
    agent: "Emily",
    kind: "reply-draft",
    tier: 2,
    subject: { realm: "Email", objectId: "e-1" },
    payload: {},
    editedPayload: null,
    rationale: "why",
    evidence: [],
    status: "pending",
    decision: null,
    createdAt: new Date(NOW - HOUR).toISOString(),
    decidedAt: null,
    holdUntil: null,
    expiresAt: null,
    question: null,
    amendments: [],
    invocationStatus: "done",
    claimedAt: null,
    ...partial,
  };
}

describe("the two clocks run OPPOSITE ways", () => {
  const p = proposal({
    id: "p",
    createdAt: new Date(NOW - 2 * HOUR).toISOString(),
    expiresAt: new Date(NOW + 3 * HOUR).toISOString(),
  });

  it("waited-for GROWS as now advances — it shames the human", () => {
    const early = rowClocks(p, NOW);
    const late = rowClocks(p, NOW + 10 * MIN);
    expect(early.waitedMs).toBe(2 * HOUR);
    expect(late.waitedMs).toBe(2 * HOUR + 10 * MIN);
    expect(late.waitedMs).toBeGreaterThan(early.waitedMs);
  });

  it("expires-in SHRINKS as now advances — it shames the clock", () => {
    const early = rowClocks(p, NOW);
    const late = rowClocks(p, NOW + 10 * MIN);
    expect(early.expiresInMs).toBe(3 * HOUR);
    expect(late.expiresInMs).toBe(3 * HOUR - 10 * MIN);
    expect(late.expiresInMs!).toBeLessThan(early.expiresInMs!);
  });

  it("a proposal with no expiresAt has NO deadline clock — never an invented one", () => {
    const open = proposal({ id: "open", expiresAt: null });
    expect(rowClocks(open, NOW).expiresInMs).toBeNull();
    expect(expiryLabel(null)).toBe("no deadline");
  });

  it("past the deadline the label says the window PASSED, not a negative count", () => {
    expect(expiryLabel(-5 * MIN)).toBe("decision window passed 5m 0s ago");
  });
});

describe("expiresAt and holdUntil are DIFFERENT clocks (s07 §T0)", () => {
  // The bug this file exists to prevent: one row carrying BOTH timestamps, and
  // each clock must come from its own field or not at all.
  const both = {
    expiresAt: new Date(NOW + 6 * DAY).toISOString(),
    holdUntil: new Date(NOW + 3 * MIN).toISOString(),
  };

  it("a HELD row renders the retraction window and NOT the decision deadline", () => {
    const held = proposal({ id: "h", status: "held", decidedAt: new Date(NOW - 2 * MIN).toISOString(), ...both });
    const clocks = rowClocks(held, NOW);
    expect(clocks.holdRemainingMs).toBe(3 * MIN); // from holdUntil…
    expect(clocks.expiresInMs).toBeNull(); // …never from expiresAt
  });

  it("a PENDING row renders the decision deadline and NOT a retraction window", () => {
    const pending = proposal({ id: "p", status: "pending", ...both });
    const clocks = rowClocks(pending, NOW);
    expect(clocks.expiresInMs).toBe(6 * DAY);
    expect(clocks.holdRemainingMs).toBeNull();
  });

  it("the hold label never claims egress — before OR after the window lapses", () => {
    // Committing out of the hold tray is s03.D T2 and does not exist, so both
    // phases must say nothing has been sent rather than implying a send.
    expect(holdLabel(3 * MIN)).toContain("nothing has been sent");
    expect(holdLabel(-1)).toContain("still not sent");
    expect(holdLabel(-1)).toContain("s03.D T2");
  });
});

describe("waited-for freezes when it stops sitting on the human", () => {
  it("freezes at decidedAt for a decided row — history stops accruing shame", () => {
    const decided = proposal({
      id: "d",
      status: "approved",
      createdAt: new Date(NOW - 5 * HOUR).toISOString(),
      decidedAt: new Date(NOW - 2 * HOUR).toISOString(),
    });
    expect(waitedForMs(decided, NOW)).toBe(3 * HOUR);
    expect(waitedForMs(decided, NOW + DAY)).toBe(3 * HOUR); // frozen
  });

  it("freezes at expiresAt for an EXPIRED row — the sweep writes no decidedAt", () => {
    const expired = proposal({
      id: "x",
      status: "expired",
      createdAt: new Date(NOW - 9 * DAY).toISOString(),
      decidedAt: null,
      expiresAt: new Date(NOW - 2 * DAY).toISOString(),
    });
    expect(waitedForMs(expired, NOW)).toBe(7 * DAY);
    expect(waitedForMs(expired, NOW + DAY)).toBe(7 * DAY);
  });

  it("labels present tense while pending, past tense after", () => {
    const p = proposal({ id: "p", createdAt: new Date(NOW - 2 * HOUR).toISOString() });
    expect(waitedLabel(p, rowClocks(p, NOW))).toBe("waiting on you 2h 0m");
    const done = proposal({ ...p, id: "q", status: "approved", decidedAt: new Date(NOW).toISOString() });
    expect(waitedLabel(done, rowClocks(done, NOW))).toBe("sat with you 2h 0m");
  });
});

describe("urgency", () => {
  it("flags near-expiry under the one-hour line, on the expiry clock only", () => {
    const near = proposal({ id: "n", expiresAt: new Date(NOW + 35 * MIN).toISOString() });
    const far = proposal({ id: "f", expiresAt: new Date(NOW + NEAR_EXPIRY_MS + MIN).toISOString() });
    expect(isNearExpiry(rowClocks(near, NOW))).toBe(true);
    expect(isNearExpiry(rowClocks(far, NOW))).toBe(false);
    // A held row's shrinking hold window is NOT expiry urgency.
    const held = proposal({ id: "h", status: "held", holdUntil: new Date(NOW + MIN).toISOString() });
    expect(isNearExpiry(rowClocks(held, NOW))).toBe(false);
  });
});

describe("orderQueue — pending first, by urgency", () => {
  const rows: ActionProposal[] = [
    proposal({ id: "expired", status: "expired", decidedAt: null, expiresAt: new Date(NOW - DAY).toISOString() }),
    proposal({ id: "held", status: "held", decidedAt: new Date(NOW - 2 * MIN).toISOString(), holdUntil: new Date(NOW + 3 * MIN).toISOString() }),
    proposal({ id: "no-deadline", createdAt: new Date(NOW - 3 * DAY).toISOString(), expiresAt: null }),
    proposal({ id: "soon", expiresAt: new Date(NOW + 35 * MIN).toISOString() }),
    proposal({ id: "later", expiresAt: new Date(NOW + 2 * DAY).toISOString() }),
    proposal({ id: "approved", status: "approved", decidedAt: new Date(NOW - 26 * HOUR).toISOString() }),
    proposal({ id: "rejected", status: "rejected", decidedAt: new Date(NOW - 2 * DAY).toISOString() }),
    // s10 T3: waiting on the AGENT — after everything decidable, before the
    // hold tray; expiresAt is null because the server banked it (paused).
    proposal({ id: "asked-late", status: "info-requested", createdAt: new Date(NOW - HOUR).toISOString(), expiresAt: null }),
    proposal({ id: "asked-early", status: "info-requested", createdAt: new Date(NOW - 5 * HOUR).toISOString(), expiresAt: null }),
  ];

  it("orders pending by soonest deadline, deadline-less last; then info-requested oldest-ask first; then held; then history newest-decided first", () => {
    expect(orderQueue(rows).map((p) => p.id)).toEqual([
      "soon",
      "later",
      "no-deadline",
      "asked-early",
      "asked-late",
      "held",
      "approved",
      "rejected",
      "expired",
    ]);
  });

  it("breaks pending deadline ties by the longest-waiting row", () => {
    const deadline = new Date(NOW + DAY).toISOString();
    const older = proposal({ id: "older", createdAt: new Date(NOW - 2 * DAY).toISOString(), expiresAt: deadline });
    const newer = proposal({ id: "newer", createdAt: new Date(NOW - HOUR).toISOString(), expiresAt: deadline });
    expect(orderQueue([newer, older]).map((p) => p.id)).toEqual(["older", "newer"]);
  });

  it("does not mutate its input", () => {
    const input = [...rows];
    orderQueue(input);
    expect(input.map((p) => p.id)).toEqual(rows.map((p) => p.id));
  });
});

describe("formatDuration", () => {
  it("is coarse with slack, seconds when short", () => {
    expect(formatDuration(2 * DAY + 3 * HOUR + 20 * MIN)).toBe("2d 3h");
    expect(formatDuration(3 * HOUR + 12 * MIN)).toBe("3h 12m");
    expect(formatDuration(35 * MIN)).toBe("35m");
    expect(formatDuration(4 * MIN + 10_000)).toBe("4m 10s");
    expect(formatDuration(41_000)).toBe("41s");
    expect(formatDuration(0)).toBe("0s");
  });
});
